### PR TITLE
Fix order status product sync

### LINKED ISSUE #1147: Wrong order count on product report
**Describe the bug**
The order count in product report summary number is incorrect

**To Reproduce**
View the footer of the revenue report, note the order count
View the summary number for orders on the order report
view the summary number for order count on the product report
Note it is different than the first two reports. 

**Expected behavior**
The order counts in the summary should match across the reports

**Screenshots**
screen shot 2018-12-20 at 7 36 37 pm

screen shot 2018-12-20 at 7 36 28 pm

screen shot 2018-12-20 at 7 36 14 pm

**Additional notes**

To validate, I looked at my orders list under the woocommerce menu. I counted the total number of orders week to date (starting the 16th of Dec) The total number of orders was 77, there were 6 failed orders, which brings the count down to 71. This number matches the revenue report, and the orders report. 

With that in mind, it looks like the products report is including failed orders (Orders count shows 77), where the other reports are not.

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #1147 

Fixes the order count and other stats where order statuses are not complete.  This PR:
* Refactors sync functions into data stores
* Checks order status and removes orders not in accepted order statuses (`completed`, `processing`, `on-hold`) for products, taxes, and coupon tables
* Fixes rebuild orders tool

### Before
screen shot 2018-12-21 at 5 46 06 pm

### After
screen shot 2018-12-21 at 5 44 56 pm

### Detailed test instructions:

1.  Create a completed order that has taxes and use a coupon.
2.  Note your existing stats prior to checking out this branch.
3.  Visit this completed order and change the status from `Completed` to `Cancelled`.
4.  Note the removal of stats for this order on the products, coupons, and taxes report.
5.  Also test that the `Rebuild Reports` tool works on `/wp-admin/admin.php?page=wc-status&tab=tools`

### CODE REVIEW

File: includes/class-wc-admin-api-init.php [hunk 1]
```diff
@@ -219,7 +218,7 @@ public static function filter_rest_endpoints( $endpoints ) {
 	/**
 	 * Regenerate data for reports.
 	 */
-	public static function regenrate_report_data() {
+	public static function regenerate_report_data() {
```
Comment (timmyc) [inline includes/class-wc-admin-api-init.php]: Good catch!

File: includes/data-stores/class-wc-admin-reports-coupons-data-store.php [hunk 2]
```diff
@@ -249,4 +258,48 @@ protected function get_cache_key( $params ) {
 		return 'woocommerce_' . self::TABLE_NAME . '_' . md5( wp_json_encode( $params ) );
 	}
 
+	/**
+	 * Make an entry in the wc_order_coupon_lookup table for an order.
```
Comment (timmyc) [inline includes/data-stores/class-wc-admin-reports-coupons-data-store.php]: Should this be "Make or Update an entry"

File: includes/data-stores/class-wc-admin-reports-coupons-data-store.php [hunk 3]
```diff
@@ -249,4 +258,48 @@ protected function get_cache_key( $params ) {
...
+	 *
+	 * @since 3.5.0
+	 * @param int $order_id Order ID.
+	 * @return void
+	 */
+	public static function sync_order_coupons( $order_id ) {
```
Comment (timmyc) [inline includes/data-stores/class-wc-admin-reports-coupons-data-store.php]: What do you think about some unit tests for this method?

File: includes/data-stores/class-wc-admin-reports-products-data-store.php [hunk 4]
```diff
@@ -287,4 +296,93 @@ protected function get_cache_key( $params ) {
 		return 'woocommerce_' . self::TABLE_NAME . '_' . md5( wp_json_encode( $params ) );
 	}
 
+	/**
+	 * Make an entry in the wc_admin_order_product_lookup table for an order.
```
Comment (timmyc) [inline includes/data-stores/class-wc-admin-reports-products-data-store.php]: Same comment, uhhhh, on the comment saying Make/Update. Also thoughts on unit tests here?

File: includes/data-stores/class-wc-admin-reports-taxes-data-store.php [hunk 5]
```diff
@@ -235,4 +244,49 @@ protected function normalize_order_by( $order_by ) {
 		return $order_by;
 	}
 
+	/**
+	 * Make an entry in the wc_order_tax_lookup table for an order.
```
Comment (timmyc) [inline includes/data-stores/class-wc-admin-reports-taxes-data-store.php]: Same comments here.

**Review by timmyc — COMMENTED** [review]
Great clean-up work here Josh! I'm fine with doing some follow-up PRs to add in the tests, but wanted to see if you think it would be a good time to increase the test coverage in this pretty key area of the logic.

**Review by coderkevin — COMMENTED** [review]
Looks good! Glad to see this using data stores!

File: includes/data-stores/class-wc-admin-reports-coupons-data-store.php (hunk 2 above)
Comment (joshuatf) [reply includes/data-stores/class-wc-admin-reports-coupons-data-store.php]: Agreed and updated!

File: includes/data-stores/class-wc-admin-reports-coupons-data-store.php (hunk 3 above)
Comment (joshuatf) [reply includes/data-stores/class-wc-admin-reports-coupons-data-store.php]: Sounds like a smart move.  Just to check before implementing this, are we okay fixing up existing tests to check for overall coverage?

I believe most of this code is covered by the save triggered in `tests/reports/class-wc-tests-reports-coupons.php`, but we need to amend this with order status data to fully cover that side of things.

Or do you think we need to unit test specifically against this method?

File: includes/data-stores/class-wc-admin-reports-products-data-store.php (hunk 4 above)
Comment (joshuatf) [reply includes/data-stores/class-wc-admin-reports-products-data-store.php]: Similar comment to above.  Unit tests here will need order status examples as well as refund data.

File: includes/data-stores/class-wc-admin-reports-coupons-data-store.php (hunk 3 above)
Comment (timmyc) [reply includes/data-stores/class-wc-admin-reports-coupons-data-store.php]: > I believe most of this code is covered by the save triggered in tests/reports/class-wc-tests-reports-coupons.php, but we need to amend this with order status data to fully cover that side of things.

Ah I was not aware of that - seems sufficient then.

**Review by timmyc — APPROVED** [review]
Good to go after a rebase :shipit:

File: includes/data-stores/class-wc-admin-reports-coupons-data-store.php (hunk 3 above)
Comment (joshuatf) [reply includes/data-stores/class-wc-admin-reports-coupons-data-store.php]: Opened an issue to increase coverage for those 2 points.  Probably needs to be done across all tests.  https://github.com/woocommerce/wc-admin/issues/1193

**Pull request merged**